### ISSUE TITLE
Compiler as a gradle/maven plugin

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#223]
Moved from https://github.com/kaitai-io/kaitai_struct_compiler/issues/103, original was by @jvmlet:

> Hi
> What is the best way to automate the generation of classes/structs ?
> The gradle/maven plugins seem to be the way to go... Are they exist ?
> If you bundle the compiler as executable jar it would be great and easy the integraition with build tools

**Comment by GreyCat** [comment]
> The gradle/maven plugins seem to be the way to go... Are they exist ?

Not right now. Contributions are most welcome :)

> If you bundle the compiler as executable jar it would be great and easy the integraition with build tools

I believe that it's already built and even deployed to Maven central repo as "io.kaitai / kaitai-struct-compiler" (snapshots only).

**Comment by jvmlet** [comment]
> I believe that it's already built and even deployed to Maven central repo as "io.kaitai / kaitai-struct-compiler" (snapshots only).

Sorry, but I can't find the `kaitai-struct-compiler` executable jar  anywhere

Meanwhile, for those who are interested, here is my solution (uses [gradle-download-task plugin](https://github.com/michel-kraemer/gradle-download-task) ,thanks a lot to Michael ) and should be slightly tweaked to suppot Unix system.

```gradle
apply plugin :'de.undercouch.download'
apply plugin :'java'

import de.undercouch.gradle.tasks.download.Download

def kaitaiVersion = 0.7
def generatedFolder =file("src/main/generated")
task downloadZipFile(type: Download) {
    src "https://bintray.com/kaitai-io/universal/download_file?file_path=${kaitaiVersion}/kaitai-struct-compiler-${kaitaiVersion}.zip"
    dest new File(buildDir, "kaitai/kaitai-struct-compiler-${kaitaiVersion}.zip")
}

task installKaitaiCompiler(dependsOn: downloadZipFile, type: Copy) {
    from zipTree(downloadZipFile.dest)
    into new File(buildDir, "kaitai")
}
task runKaitaiCompiler(dependsOn: installKaitaiCompiler, type: Exec) {
    workingDir new File(buildDir, "kaitai/kaitai-struct-compiler-${kaitaiVersion}/bin")
    executable 'cmd' // tweak for UNIX
    args '/c'  // tweak for UNIX
    args 'kaitai-struct-compiler.bat'  // tweak for UNIX
    args '-t','java'
    args "-d","${generatedFolder}"
    args "--java-package","my.company"
    args "${buildDir}/kaitai/kaitai-struct-compiler-0.7/formats/image/bmp.ksy" // set your's ksy here


}
sourceSets.main.java {
    srcDir "${generatedFolder}/src"
}
compileJava.dependsOn  runKaitaiCompiler
dependencies{
    compile "io.kaitai:kaitai-struct-runtime:${kaitaiVersion}"
}

```

**Comment by GreyCat** [comment]
@jvmlet The executable jar is published only as snapshots, thus basically, here it is: https://oss.sonatype.org/content/repositories/snapshots/io/kaitai/kaitai-struct-compiler_2.11/

Or, if you prefer, [direct link to jar](https://oss.sonatype.org/content/repositories/snapshots/io/kaitai/kaitai-struct-compiler_2.11/0.8-SNAPSHOT/kaitai-struct-compiler_2.11-0.8-SNAPSHOT.jar). Probably it would be easier to launch a class in jar:

* [io.kaitai.struct.JavaMain#main](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/jvm/src/main/scala/io/kaitai/struct/JavaMain.scala#L195) — that's entry point that parses CLI arguments, or
* [io.kaitai.struct.JavaMain#run](https://github.com/kaitai-io/kaitai_struct_compiler/blob/master/jvm/src/main/scala/io/kaitai/struct/JavaMain.scala#L209-L230) —that's entry point that does just the compiling given the CLIConfig object that specifies all the options

**Comment by jvmlet** [comment]
Thanks, when are you planing to release 0.8 ?

**Comment by GreyCat** [comment]
To be frank, it was supposed to ship in July, 2017, but we're terribly falling behind. There's still [a few unsolved issues](https://github.com/kaitai-io/kaitai_struct/milestone/5), some of them are considered critical to be in 0.8. So, no hard dates are available yet :(

**Comment by daviderickson** [comment]
@GreyCat Is there a reason that there isn't a kaitai-struct-compiler full release to maven for 0.7?

**Comment by GreyCat** [comment]
I believe no one asked/needed that. Technically it should be doable, although I've never tried publishing an sbt-built artifact into a public mvn repo.

**Comment by daviderickson** [comment]
That would be really helpful.  You might also consider using a build tool that makes pushing to Maven easy, since virtually everyone developing on the JVM uses that as the main library source.

Thanks!

**Comment by GreyCat** [comment]
@daviderickson Could you recommend something to making Maven pushing easier, and how exactly would that "easy" look like? Right now, publishing with sbt takes a single `sbt publish` command, but the rest of the painful process is Sonatype's web UI :(

**Comment by daviderickson** [comment]
I've never used SBT, but we build all our internal projects with Maven and publish with the command line, never needing to go Sonatype's UI.  Ours may be configured differently from maven central though, they may require it, I don't know.

**Comment by jvisser** [comment]
Any progress on this? I would like to call the compiler during (maven) build process to generate sources from ksy before compiling and packaging. Final release of the compiler in maven central would make this a lot simpeler and more robust (no snapshot and custom maven repo required).

**Comment by KOLANICH** [comment]
#466 is related.

**Comment by GreyCat** [comment]
@valery1707 has started a project here, actually: https://github.com/valery1707/kaitai-maven-plugin

I'll try to get official releases of 0.8 pushed into the permanent repos soon.

**Comment by valery1707** [comment]
@jvmlet, @jvisser  you can use my project https://github.com/valery1707/kaitai-maven-plugin to build Kaitai templates into Java-classes. Version `0.1.1` already use Kaitai `0.8` by default but you can configure any other version.
I have demo project in https://github.com/valery1707/kaitai-java-demo

**Comment by jvmlet** [comment]
This is how I'm integrating kaitai compiler into the buld :
```gradle
configurations{
    kaitai
}

def generatedFolder =file("src/main/generated")

def kaitaiVersion="0.8-SNAPSHOT"
dependencies{
    compile "io.kaitai:kaitai-struct-runtime:${kaitaiVersion}"
    kaitai "io.kaitai:kaitai-struct-compiler_2.11:${kaitaiVersion}"
}
task runKaitaiCompiler(type: JavaExec) {

    doFirst{
        classpath configurations.kaitai.asFileTree
    }
    main  "io.kaitai.struct.JavaMain"
    args '-t','java'
    args "-d","${generatedFolder}"
    args "--java-package","com.my.company"

    Arrays.asList(new File("${projectDir}/src/main/resources").listFiles(new FilenameFilter() {
        boolean accept(File dir, String name) {
            return name.endsWith(".ksy")
        }
    })).stream().map(new Function<File,String>(){
        @Override
        String apply(File file) {
            return file.getAbsolutePath();
        }
    }).each {args it}

}
sourceSets.main.java {
    srcDir "${generatedFolder}/src"
}
compileJava.dependsOn  runKaitaiCompiler
```

**Comment by valery1707** [comment]
@jvmlet I planning to create a gradle plugin soon

**Comment by valery1707** [comment]
@jvmlet, @jvisser, @YangFan789 [kaitai-gradle-plugin](https://github.com/valery1707/kaitai-gradle-plugin) was released.
You can see documentation in project home or use [demo project](https://github.com/valery1707/kaitai-java-demo/tree/master/gradle-simple) directly.

**Comment by valery1707** [comment]
@GreyCat can you hint me how I can add information about [kaitai-maven-plugin](https://github.com/valery1707/kaitai-maven-plugin)/[kaitai-gradle-plugin](https://github.com/valery1707/kaitai-gradle-plugin) into [Kaitai documentation](http://doc.kaitai.io/lang_java.html)?
The [source file](https://github.com/kaitai-io/doc-kaitai-io.github.io/blob/master/lang_java.html) for this page looks generated. If it's true - where I can find the sources for generating?

**Comment by koczkatamas** [comment]
Here is the source: https://github.com/kaitai-io/kaitai_struct_doc/blob/master/lang_java.adoc

**Comment by prasannabalaraman** [comment]
I am getting the following exception while running kaitai task in gradle.

Execution failed for task ':kaitai'.

name.valery1707.kaitai.KaitaiException: Fail to download 'https://dl.bintray.com/kaitai-io/universal/0.8/kaitai-struct-compiler-0.8.zip' into '/build/tmp/kaitai-cache/kaitai-struct-compiler-0.8.zip.tmp'

I believe this is because jcenter is no longer available. Please let me know how to overcome this.

**Comment by generalmimon** [comment]
@prasannabalaraman:

> I believe this is because jcenter is no longer available. Please let me know how to overcome this.

It seems that you're able to override the URL of the downloaded ZIP archive (see also https://github.com/valery1707/kaitai-gradle-plugin/tree/6837013#plugin-parameters). This has also been suggested by @valery1707 in this issue - https://github.com/valery1707/kaitai-gradle-plugin/issues/20#issuecomment-852209344. For example:

[Artur51](https://github.com/Artur51) / [**kaitai-psd-read-font-settings**](https://github.com/Artur51/kaitai-psd-read-font-settings) / [`build.gradle:20-26`](https://github.com/Artur51/kaitai-psd-read-font-settings/blob/90bf3650dbe4c424c6864375aeb17ad0eb00a78f/build.gradle#L20-L26)

```gradle
//https://github.com/valery1707/kaitai-gradle-plugin
kaitai {
    packageName = 'com.gen.kaitai.psd'
    output = new File('src/main/java')
    version = '0.9'
    url = new URL("https://github.com/kaitai-io/kaitai_struct_compiler/releases/download/0.9/kaitai-struct-compiler-0.9.zip")
}
```

**Comment by generalmimon** [comment]
But from the perspective of https://github.com/valery1707/kaitai-gradle-plugin, I don't fully agree with the final @valery1707's word in https://github.com/valery1707/kaitai-gradle-plugin/issues/20#issuecomment-852209344 and closing that issue as `wontfix`.

The latest released [`kaitai-gradle-plugin:0.1.3`](https://search.maven.org/artifact/name.valery1707.kaitai/kaitai-gradle-plugin/0.1.3/jar) uses [`kaitai-maven-plugin:0.1.6`](https://search.maven.org/artifact/name.valery1707.kaitai/kaitai-maven-plugin/0.1.6/maven-plugin) (which is the latest release of `kaitai-maven-plugin`), which still has the wrong Bintray URL hardcoded as the default:

https://github.com/valery1707/kaitai-maven-plugin/blob/v0.1.6/src/main/java/name/valery1707/kaitai/KaitaiUtils.java#L362

```java
	private static final String URL_FORMAT = "https://dl.bintray.com/kaitai-io/universal/%s/kaitai-struct-compiler-%s.zip";
```

It was fixed in https://github.com/valery1707/kaitai-maven-plugin/commit/22fe32252c15d13eac97ac7a469ff451ac65fb1b, but that was only after the 0.1.6 release (obviously, because https://github.com/valery1707/kaitai-maven-plugin/commit/22fe32252c15d13eac97ac7a469ff451ac65fb1b is from Jul 1, 2021 and 0.1.6 was released on Dec 4, 2019):

```diff
-	private static final String URL_FORMAT = "https://dl.bintray.com/kaitai-io/universal/%s/kaitai-struct-compiler-%s.zip";
+	private static final String URL_FORMAT = "https://github.com/kaitai-io/kaitai_struct_compiler/releases/download/%s/kaitai-struct-compiler-%s.zip";
```

So it appears to me that ideally there should be a new version of `kaitai-maven-plugin` released and then also `kaitai-gradle-plugin` with the updated `kaitai-maven-plugin` dependency to change this default URL. Because right now, every single user of `kaitai-gradle-plugin`/`kaitai-maven-plugin` has to override it, because the default never works.

**Comment by prasannabalaraman** [comment]
Thanks @generalmimon. I will overriding the url parameter

**Comment by prasannabalaraman** [comment]
@generalmimon I have this in my build.gradle 

For some reason it is not generating the java files but just creates the folder in build/generated/sources/kaitai/java/main.

All my .ksy files are at src/main/java/resources/kaitai folder

buildscript {
    dependencies {
        classpath("name.valery1707.kaitai:kaitai-gradle-plugin:0.1.3")
    }
}

apply plugin: 'name.valery1707.kaitai'
dependencies {
    implementation 'io.kaitai:kaitai-struct-runtime:0.10'
}

    kaitai {
        packageName = 'com.package.name.kaitai'
        output = new File('build/generated/sources/kaitai/java/main')
        version = '0.9'
        url = new URL("https://github.com/kaitai-io/kaitai_struct_compiler/releases/download/0.9/kaitai-struct-compiler-0.9.zip")
    }

    test {
        testLogging {
            events("passed", "skipped", "failed")
        }
    }

Note: I am able to generate java files based on the approach from @jvmlet with iterating .ksy files.

**Comment by generalmimon** [comment]
@prasannabalaraman:

> All my .ksy files are at src/main/java/resources/kaitai folder

According to https://github.com/valery1707/kaitai-gradle-plugin#plugin-parameters, they are expected in `src/main/resources/kaitai` (not <code>src/main/<strong><del>java/</strong></del>resources/kaitai</code> as you write):

> ## Plugin parameters
> 
> Name | Type | Since | Description
> -- | -- | -- | --
> ... | ... | ... | ...
> sourceDirectory | java.io.File | 0.1.0 | Source directory with [Kaitai Struct language](http://formats.kaitai.io/) files.<br><br>**Default**: src/main/resources/kaitai

---

Also, I would suggest using the latest `kaitai-struct-compiler` 0.10 - it has several improvements over 0.9 on which you'd miss out:

```gradle
        version = '0.10'
        url = new URL("https://github.com/kaitai-io/kaitai_struct_compiler/releases/download/0.10/kaitai-struct-compiler-0.10.zip")
```

**Comment by prasannabalaraman** [comment]
@generalmimon 

Sorry, all my .ksy files are at src/main/resources/kaitai.

I changed the version to 0.10 and gradle is fetching the .ksy files and throwing bunch of warnings like

warning: use `len_hashes` instead of `hash_size`, given that it's only used as a byte size of `hashes` 

and it is not generating the java files. I downloaded mach_o.ksy from https://github.com/kaitai-io/kaitai_struct_formats/blob/master/executable/mach_o.ksy

Appreciate your help.

**Comment by generalmimon** [comment]
@prasannabalaraman Are you using the format class somewhere, i.e. importing `com.package.name.kaitai.MachO` and using the `MachO` class?

**Comment by generalmimon** [comment]
Hmm, but given that the compiler was apparently invoked, as you write:

> I changed the version to 0.10 and gradle is fetching the .ksy files and throwing bunch of warnings like
> 
> warning: use `len_hashes` instead of `hash_size`, given that it's only used as a byte size of `hashes`

the Java files should really be somewhere - that's weird...

**Comment by generalmimon** [comment]
@prasannabalaraman 

> gradle is fetching the .ksy files and throwing bunch of warnings like

Do you see any `error:` there in the compiler output? Errors should be at the beginning. Warnings are fine, they do not block compilation.

**Comment by prasannabalaraman** [comment]
@generalmimon 

Yes I will be using MachO.java in my application (that is the next step).

I could not figure out the generated java files anywhere. 

I assume it should be at this location **output = new File('build/generated/sources/kaitai/java/main')** It is creating the folder but not the files underneath.

Probably I will go with @jvmlet approach for now

**Comment by prasannabalaraman** [comment]
> error

I do not see any error in the logs. Also the build is success

12:33:45 PM: Executing task 'build --stacktrace -x test'...


> Task :kaitai
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/cs_blob/types/code_directory/seq/7/id:
warning: use `len_hashes` instead of `hash_size`, given that it's only used as a byte size of `hashes` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/cs_blob/types/data/seq/0/id:
warning: use `len_value` instead of `length`, given that it's only used as a byte size of `value` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/cs_blob/types/super_blob/seq/0/id:
warning: use `num_blobs` instead of `count`, given that it's only used as repeat count of `blobs` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/cs_blob/types/requirements/seq/0/id:
warning: use `num_items` instead of `count`, given that it's only used as repeat count of `items` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/build_version_command/seq/3/id:
warning: use `num_tools` instead of `ntools`, given that it's only used as repeat count of `tools` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/segment_command_64/seq/7/id:
warning: use `num_sections` instea
d of `nsects`, given that it's only used as repeat count of `sections` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dysymtab_command/seq/13/id:
warning: use `num_indirect_symbols` instead of `n_indirect_syms`, given that it's only used as repeat count of `indirect_symbols` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/code_signature_command/seq/1/id:
warning: use `len_code_signature` instead of `data_size`, given that it's only used as a byte size of `code_signature` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/seq/1/id:
warning: use `len_rebase` instead of `rebase_size`, given that it's only used as a byte size of `rebase` (see https://doc.kaitai.io/ksy_style_guide.ht
ml#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/seq/3/id:
warning: use `len_bind` instead of `bind_size`, given that it's only used as a byte size of `bind` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/seq/7/id:
warning: use `len_lazy_bind` instead of `lazy_bind_size`, given that it's only used as a byte size of `lazy_bind` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/seq/9/id:
warning: use `len_exports` instead of `export_size`, given that it's only used as a byte size of `exports` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/types/export_node/seq/1/id:
warning: use `num_children` instead of `children_count`, given that it's only used as repeat count of `children` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/segment_command/seq/7/id:
warning: use `num_sections` instead of `nsects`, given that it's only used as repeat count of `sections` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/segment_command/types/section/seq/3/id:
warning: use `len_data` instead of `size`, given that it's only used as a byte size of `data` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/symtab_command/seq/3/id:
warning: use `len_strs` instead of `str_size`, given that it's only used as a byte size of `strs` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/symtab_command/seq/1/id:
warning: use `num_symbols` instead of `n_syms`, given that it's only used as repeat count of `symbols` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/cs_blob/types/code_directory/seq/7/id:
warning: use `len_hashes` instead of `hash_size`, given that it's only used as a byte size of `hashes` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/cs_blob/types/data/seq/0/id:
warning: use `len_value` instead of `length`, given that it's only used as a byte size of `value` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/cs_blob/types/super_blob/seq/0/id:
warning: use `num_blobs` inst
ead of `count`, given that it's only used as repeat count of `blobs` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/cs_blob/types/requirements/seq/0/id:
warning: use `num_items` instead of `count`, given that it's only used as repeat count of `items` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/build_version_command/seq/3/id:
warning: use `num_tools` instead of `ntools`, given that it's only used as repeat count of `tools` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/segment_command_64/seq/7/id:
warning: use `num_sections` instead of `nsects`, given that it's only used as repeat count of `sections` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworksp
ace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dysymtab_command/seq/13/id:
warning: use `num_indirect_symbols` instead of `n_indirect_syms`, given that it's only used as repeat count of `indirect_symbols` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/code_signature_command/seq/1/id:
warning: use `len_code_signature` instead of `data_size`, given that it's only used as a byte size of `code_signature` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/seq/1/id:
warning: use `len_rebase` instead of `rebase_size`, given that it's only used as a byte size of `rebase` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/seq/3/id:
warning: use `len_bind`
instead of `bind_size`, given that it's only used as a byte size of `bind` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/seq/7/id:
warning: use `len_lazy_bind` instead of `lazy_bind_size`, given that it's only used as a byte size of `lazy_bind` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/seq/9/id:
warning: use `len_exports` instead of `export_size`, given that it's only used as a byte size of `exports` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/dyld_info_command/types/export_node/seq/1/id:
warning: use `num_children` instead of `children_count`, given that it's only used as repeat count of `children` (see https://doc.kaitai.io/ksy_style_guide.html
#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/segment_command/seq/7/id:
warning: use `num_sections` instead of `nsects`, given that it's only used as repeat count of `sections` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/segment_command/types/section/seq/3/id:
warning: use `len_data` instead of `size`, given that it's only used as a byte size of `data` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/symtab_command/seq/3/id:
warning: use `len_strs` instead of `str_size`, given that it's only used as a byte size of `strs` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/mach_o.ksy: /types/symtab_command/seq/1/id:
warning: use `num_symbols` instead of
`n_syms`, given that it's only used as repeat count of `symbols` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/elf.ksy: /types/endian_elf/seq/8/id:
warning: use `len_program_headers` instead of `program_header_entry_size`, given that it's only used as a byte size of `program_headers` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/elf.ksy: /types/endian_elf/seq/10/id:
warning: use `len_section_headers` instead of `section_header_entry_size`, given that it's only used as a byte size of `section_headers` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/elf.ksy: /types/endian_elf/seq/9/id:
warning: use `num_program_headers` instead of `qty_program_header`, given that it's only used as repeat count of `program_headers` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)
/Users/pbalaraman/intellijworkspace/scanner_common/src/main/resources/kaitai/elf.ksy: /types/endian_elf/seq/11/id:
warning: use `num_section_headers` instead of `qty_section_header`, given that it's only used as repeat count of `section_headers` (see https://doc.kaitai.io/ksy_style_guide.html#attr-id)

> Task :compileJava
> Task :processResources
> Task :classes
> Task :jar
> Task :assemble
> Task :check
> Task :build

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.5.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 7s
4 actionable tasks: 4 executed
12:33:53 PM: Task execution finished 'build --stacktrace -x test'.

**Comment by generalmimon** [comment]
@prasannabalaraman Try specifying an absolute path in the `output` parameter.

**Comment by prasannabalaraman** [comment]
That works and I see the java files now.

I am getting an exception after adding generated classes in gradle

    sourceSets {
        main {
            java.srcDirs += 'build/generated/sources/kaitai/java/main'
        }
    }

/Users/pbalaraman/intellijworkspace/scanner_common/build/generated/sources/kaitai/java/main/com/bluecedar/common/scanner/kaitai/MachO.java:2050: error: incompatible types: possible lossy conversion from long to int
                        switch (id()) {
                               ^
Not sure **https://github.com/kaitai-io/kaitai_struct_formats/blob/master/executable/mach_o.ksy** is the correct version?

**Comment by generalmimon** [comment]
> /Users/pbalaraman/intellijworkspace/scanner_common/build/generated/sources/kaitai/java/main/com/bluecedar/common/scanner/kaitai/MachO.java:2050: error: incompatible types: possible lossy conversion from long to int
> switch (id()) {
> ^
> Not sure **https://github.com/kaitai-io/kaitai_struct_formats/blob/master/executable/mach_o.ksy** is the correct version?

It is the latest available version of the spec, but unfortunately, the compiler doesn't generate Java code that `javac` would be happy with 😞

Can you check if the following .ksy patch fixes the issue?

[`executable/mach_o.ksy:468-476`](https://github.com/kaitai-io/kaitai_struct_formats/blob/6c7ce1f88d6dc02c24b651abfd92ecb701e1aa35/executable/mach_o.ksy#L468-L476)

```diff
               - id: id
                 type: u4
               - id: body
                 size: length - 4
                 if: length > 0
                 type:
-                  switch-on: id
+                  switch-on: id.as<s4>
                   cases:
                     0: cie
```
